File: go/bit/67_add_binary.go
package bit

import "strconv"

// addBinary 二进制求和
func addBinary(a, b string) string {

	ans := ""

	// 进位
	carry := 0

	lenA, lenB := len(a), len(b)
	n := max(lenA, lenB)

	for i := 0; i < n; i++ {
		if i < lenA {
			// 字符转成数字
			carry += int(a[lenA-i-1] - '0')
		}

		if i < lenB {
			carry += int(b[lenB-i-1] - '0')
		}
		// 求余之后是各位数字
		ans = strconv.Itoa(carry%2) + ans
		// 求商是进位数字
		carry /= 2
	}

	if carry > 0 {
		ans = "1" + ans
	}

	return ans
}

File: rust/src/bit/_67_add_binary.rs
/// 二进制求和
pub fn add_binary(a: String, b: String) -> String {
    let mut ans = String::new();

    let (a, b): (Vec<_>, Vec<_>) = (a.chars().collect(), b.chars().collect());

    let (len_a, len_b) = (a.len(), b.len());

    let n = len_a.max(len_b);

    let mut carry = 0;

    for i in 0..n {
        if i < len_a {
            let digit = a[len_a - i - 1].to_digit(10).unwrap();
            carry += digit;
        }

        if i < len_b {
            let digit = b[len_b - i - 1].to_digit(10).unwrap();
            carry += digit;
        }
        // 加法运算余数放到前面
        ans = format!("{}{}", carry % 2, ans);
        carry /= 2;
    }

    if carry > 0 {
        ans = format!("{}{}", 1, ans);
    }

    ans
}
